Sources/NIOPosix/NonBlockingFileIO.swift

```diff
@@ -712,3 +730,379 @@ public struct NIODirectoryEntry: Hashable {
+                try Posix.ftruncate(descriptor: descriptor, size: off_t(size))
+            }
+        }
+    }
+
+    /// Returns the length of the file associated with `fileHandle`.
+    ///
+    /// - parameters:
+    ///   - fileHandle: The `NIOFileHandle` to read from.
+    @available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+    public func readFileSize(fileHandle: NIOFileHandle) async throws -> Int64 {
+        return try await self.threadPool.runIfActive {
+            return try fileHandle.withUnsafeFileDescriptor { descriptor in
+                let curr = try Posix.lseek(descriptor: descriptor, offset: 0, whence: SEEK_CUR)
+                let eof = try Posix.lseek(descriptor: descriptor, offset: 0, whence: SEEK_END)
+                try Posix.lseek(descriptor: descriptor, offset: curr, whence: SEEK_SET)
+                return Int64(eof)
+            }
+        }
+    }
+
+    /// Write `buffer` to `fileHandle` in ``NonBlockingFileIO``'s private thread pool.
+    ///
+    /// - parameters:
+    ///   - fileHandle: The `NIOFileHandle` to write to.
+    ///   - buffer: The `ByteBuffer` to write.
+    @available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+    public func write(fileHandle: NIOFileHandle,
+                      buffer: ByteBuffer) async throws {
+        return try await self.write0(fileHandle: fileHandle, toOffset: nil, buffer: buffer)
+    }
+
+    /// Write `buffer` starting from `toOffset` to `fileHandle` in ``NonBlockingFileIO``'s private thread pool.
+    ///
+    /// - parameters:
+    ///   - fileHandle: The `NIOFileHandle` to write to.
+    ///   - toOffset: The file offset to write to.
+    ///   - buffer: The `ByteBuffer` to write.
+    @available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+    public func write(fileHandle: NIOFileHandle,
+                      toOffset: Int64,
+                      buffer: ByteBuffer) async throws {
+        return try await self.write0(fileHandle: fileHandle, toOffset: toOffset, buffer: buffer)
+    }
+
+    @available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+    private func write0(fileHandle: NIOFileHandle,
+                        toOffset: Int64?,
+                        buffer: ByteBuffer) async throws {
+        let byteCount = buffer.readableBytes
+
+        guard byteCount > 0 else {
+            return
+        }
+
+        return try await self.threadPool.runIfActive {
+            try self.writeSync(fileHandle: fileHandle, byteCount: byteCount, toOffset: toOffset, buffer: buffer)
+        }
+    }
+
+    /// Open file at `path` on a private thread pool, run an operation given the file handle and region and then close the file handle.
+    ///
+    /// The open file operation runs on a private thread pool.
+    ///    
+    /// - note: The reason this provides the `NIOFileHandle` and the `FileRegion` is that both the opening of a file as well as 
+    /// the querying of its size are blocking.
```

Comment:
Do we actually need to provide both in the closure? `FileRegion` has a `fileHandle` property which is `public` so it's available if users need it. `withFileRegion` would be a more direct analogue to `withFileHandle` than `withOpenFile` as well.

Reply:
That's a nice idea. This is now implemented